Feature: Add include option to unarchive

sijis opened the pull request:
##### SUMMARY

This should allow users to extract specific files from an archive as desired.

Fixes #16130, #27081.

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Feature Pull Request

##### COMPONENT NAME
<!--- Name of the module, plugin, module or task -->
unarchive

##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes below -->
```
ansible 2.6.0a1 (feature/unarchive/add_include_option 70f23af274) last updated 2018/05/22 01:18:14 (GMT -500)
  config file = /home/saviles/data/git/github/ansible/ansible.cfg
  configured module search path = ['/home/saviles/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /home/saviles/data/git/github/ansible/lib/ansible
  executable location = /home/saviles/data/git/github/ansible/bin/ansible
  python version = 3.6.5 (default, Mar 29 2018, 18:20:46) [GCC 8.0.1 20180317 (Red Hat 8.0.1-0.19)]

```


##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful. -->

**Playbook**
```
- hosts: localhost
  vars:
    issue: 16130
    src_dir: "files/{{ issue }}"
    dst_dir: "/tmp/{{ issue }}"
  tasks:
    - name: Create directory
      file:
        state: directory
        path: "{{item}}"
      with_items:
        - "{{src_dir}}"
        - "{{dst_dir}}"
        - "{{dst_dir}}-tar"
    - name: Create files
      shell: "touch {{src_dir}}/{include,exclude}-{0..9}.txt"

    - name: Create archive file
      archive:
        format: "zip"
        path: "{{src_dir}}/*"
        dest: "{{src_dir}}.zip"
    - name: Unpack zip file
      unarchive:
        src: "{{src_dir}}.zip"
        dest: "{{dst_dir}}"
        remote_src: yes
        include:
          - 'include-2.txt'
      register: unzipped

    - name: Create archive file
      archive:
        format: "tar"
        path: "{{src_dir}}/*"
        dest: "{{src_dir}}.tar"
    - name: Unpack zip file
      unarchive:
        src: "{{src_dir}}.tar"
        dest: "{{dst_dir}}-tar"
        # remote_src: yes
        include:
          - 'include-4.txt'
```

<!--- Paste verbatim command output below, e.g. before and after your change -->
**Output**
```
$ tree /tmp/16130*
/tmp/16130
└── include-2.txt
/tmp/16130-tar
└── include-4.txt

0 directories, 2 files
```

sijis committed:
feature: Add include option to unarchive

This should allow users to extract specific files from an archive as
desired.

Fixes #16130, #27081.

sijis commented:
The only thing left is adding integration tests.

sijis commented:
Added tests for zip archive. Need to complete for gz.

sijis commented:
Added fix to tar files including tests.
I also had to rebase, as the `excludes` feature fix PR was added.

sijis commented:
Something to ponder as I was rebasing.

I wonder if his option `include` should be mutually exclusive with the `exclude` option.
We could support both but we'd have handle situations where someone 'includes' something, then excludes a set of that. We'd also have to handle the opposite behavior.
Conversely having them excludes would make thing simpler *and* a user is likely to use one of the two approaches (either include or exclude files).

cacack commented:
@sijis Any word on this?  Writing a playbook right now that would greatly benefit from being able to whitelist the files to extract...

tchernomax commented:
LGTM

though I would have use `fnmatch.fnmatch(member, include)` (as in exclude), but I can understand that "include" should not use glob.

s-hertel commented on lib/ansible/modules/files/unarchive.py:
```diff
@@ -802,6 +822,7 @@ def main():
             list_files=dict(type='bool', default=False),
             keep_newer=dict(type='bool', default=False),
             exclude=dict(type='list', default=[]),
+            include=dict(type='list', default=[]),
```
Sorry for taking a while to review this one! Do you mind rebasing this, updating the version added, and adding a changelog?

This option should probably be mutually exclusive with `exclude` - i.e. `AnsibleModule(..., mutually_exclusive=[('exclude', 'include',)])`

samdoran committed:
Rebase and make a few minor changes

samdoran committed:
Add changelog

samdoran committed:
Improve tests

- move to separate tasks file
- change assertions to check for exactly one file
- use remote_tmp_dir for output dir

samdoran committed:
Make exclude and include mutually exclusive

samdoran committed:
Don't remove files needed by other tasks

samdoran committed:
Fix sanity tests

samdoran committed:
Improve feature documentation

samdoran committed:
Skip tests that use map() on CentOS 6

samdoran committed:
Use fnmatch on include for zip archives

This matches the behavior of exclude

samdoran committed:
Fix bad merge

sijis commented:
@samdoran Thanks for rebasing and moving this along.

samdoran commented:
@sijis Sorry it took us so long to merge this. Thank you for your contribution.
